### ISSUE TITLE
MongoDB client based app fails at building native image due to missing class 'jnr.unixsocket.UnixSocket'

### ISSUE THREAD

**Issue by mswiderski** [quarkusio/quarkus#24582]
### Describe the bug

When building an application that uses `mongodb-client` extension into native image the build fails with following exception

````
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Running Quarkus native-image plugin on GraalVM 21.3.1 Java 11 CE (Java Version 11.0.14+9-jvmci-21.3-b09)
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] docker run --env LANG=C --rm -v /Users/mswiderski/Downloads/quarkus-quickstarts-main-2/mongodb-quickstart/target/mongodb-quickstart-1.0.0-SNAPSHOT-native-image-source-jar:/project:z --name build-native-ExZeM quay.io/quarkus/ubi-quarkus-native-image:21.3-java11 -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=3 -J-Duser.language=en -J-Duser.country=PL -J-Dfile.encoding=UTF-8 -H:-ParseOnce -J--add-exports=java.security.jgss/sun.security.krb5=ALL-UNNAMED -J--add-opens=java.base/java.text=ALL-UNNAMED --report-unsupported-elements-at-runtime -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -J-Djava.awt.headless=true -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http,https -H:-UseServiceLoaderFeature -H:+StackTrace mongodb-quickstart-1.0.0-SNAPSHOT-runner -jar mongodb-quickstart-1.0.0-SNAPSHOT-runner.jar
[mongodb-quickstart-1.0.0-SNAPSHOT-runner:25]    classlist:   6,031.38 ms,  0.94 GB
[mongodb-quickstart-1.0.0-SNAPSHOT-runner:25]        setup:     963.02 ms,  0.94 GB
Fatal error:java.lang.NoClassDefFoundError: jnr/unixsocket/UnixSocket
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166)
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2309)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleDeletedClass(AnnotationSubstitutionProcessor.java:525)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:306)
	at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:266)
	at com.oracle.svm.hosted.NativeImageGenerator.createDeclarativeSubstitutionProcessor(NativeImageGenerator.java:912)
	at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:839)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:527)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:488)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:403)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:569)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:122)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:599)
Caused by: java.lang.ClassNotFoundException: jnr.unixsocket.UnixSocket
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:476)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 14 more
[mongodb-quickstart-1.0.0-SNAPSHOT-runner:25]      [total]:   7,107.81 ms,  0.94 GB
# Printing build artifacts to: /project/mongodb-quickstart-1.0.0-SNAPSHOT-runner.build_artifacts.txt
Error: Image build request failed with exit status 1
````

This only happens when application uses 
````
<quarkus.native.additional-build-args>--report-unsupported-elements-at-runtime</quarkus.native.additional-build-args>
````
in the `native` maven profile.

Looks like it triggers some substitutions or something around that. 

Possible workarounds

- remove the `<quarkus.native.additional-build-args>--report-unsupported-elements-at-runtime</quarkus.native.additional-build-args>` but sometimes this is not an option
- declare `com.github.jnr:jnr-unixsocket` dependency in the project

### Expected behavior

native build image should succeed with `<quarkus.native.additional-build-args>--report-unsupported-elements-at-runtime</quarkus.native.additional-build-args>` and `mongodb-client` extension

### How to Reproduce?

1. Clone mongo db client Quickstart (https://github.com/quarkusio/quarkus-quickstarts/tree/main/mongodb-quickstart)

2.Add `<quarkus.native.additional-build-args>--report-unsupported-elements-at-runtime</quarkus.native.additional-build-args>` into properties of `native` maven profile

3. Run native build ` mvn clean package -DskipTests -Dquarkus.native.container-build=true -Dquarkus.container-image.build=true  -Pnative` 

### Output of `uname -a` or `ver`

Darwin MacBook-Pro.local 21.1.0 Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64 x86_64

### Output of `java -version`

Java version: 11.0.11, vendor: AdoptOpenJDK

### GraalVM version (if different from Java)

GraalVM 21.3.1 Java 11 CE

### Quarkus version or git rev

2.7.5.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3

**Comment by geoand** [comment]
@galderz  @zakkak any idea what's going on here?

**Comment by loicmathieu** [comment]
As a general rule of thumb, I would say that we cannot guarantee that our native image support is correct if you pass a paremeter that will modify the native image behaviour.

So if you pass `--report-unsupported-elements-at-runtime` this ca cause some part of our support to be skipped, or modify the way GraalVM native image would analyse the code and we cannot do anything about it. We already had some case where some library brings this parameter automatically (via native-image.properties) and cannot do anything that asking the library to remove it.

So my advise: remove the parameter, if it crashes at build time copy/paste the error here maybe we can help.

**Comment by zakkak** [comment]
When using `--report-unsupported-elements-at-runtime` GraalVM [registers all the reachable types, even for unsupported/deleted elements](https://github.com/oracle/graal/blob/bc6d4b892055d7147a7ec8293de3d82a89725eef/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/substitute/AnnotationSubstitutionProcessor.java#L583-L605).

In this case it fails while trying to load the reachable types of `com.mongodb.internal.connection.UnixSocketChannelStream` which Quarkus deletes in https://github.com/quarkusio/quarkus/blob/0db120a1b60addfea6982d9d31b14a17eea99906/extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/graal/MongoClientSubstitutions.java#L126-L130

`com.github.jnr:jnr-unixsocket` appears to be an (optional?) runtime dependency of `mongo-db-driver-core`, which due to the explicit deletion of `com.mongodb.internal.connection.UnixSocketChannelStream` Quarkus isn't really needed, yet `--report-unsupported-elements-at-runtime` result in it being loaded at build-time.

So to sum up, when using `--report-unsupported-elements-at-runtime` you might need to augment the classpath, but it's not worth increasing the dependencies of the default configuration to support that case.

**Comment by geoand** [comment]
Thanks for the analysis @zakkak! 

I'll close this then as it's not a Quarkus issue.

**Comment by loicmathieu** [comment]
Thanks for the analysis @zakkak, very clear !

**Comment by mswiderski** [comment]
excellent explanation @zakkak! thanks a lot for that.
